### PR TITLE
Add option to toggle enabling cron events.

### PR DESCRIPTION
**Pull request by jaredcobb**
Also:
- Refactor do_post_processing to have better reporting, better memory
management.
- Added new CLI command force_delete_objects() to remove all objects
from the database. (Useful for re-synchronizing from scratch).

### CODE REVIEW

File: inc/class-tmsc-sync.php [hunk 1]
```diff
@@ -308,45 +324,58 @@ public function complete_sync() {
...
+		if ( ! empty( $total_count->total ) ) {
+			if ( $cli ) {
+				\WP_CLI::line( sprintf( '%s posts need processing...', $total_count->total ) );
+			}
+
+			do {
+
+				$post_processing_data = $wpdb->get_results(
+					$wpdb->prepare(
+						"SELECT post_id, meta_value FROM {$wpdb->postmeta} WHERE meta_key = %s AND post_id = 436501 ORDER BY post_id LIMIT %d, %d",
```
Comment (kjbenk) [inline inc/class-tmsc-sync.php]: Is the `AND post_id = 436501` debug code?

File: inc/class-tmsc-sync.php [hunk 2]
```diff
@@ -308,45 +324,58 @@ public function complete_sync() {
...
+					$batch_position = ( $batch_number * $batch_size ) - ( $batch_size - 1 );
+					if ( $cli ) {
+						\WP_CLI::line( sprintf( 'Beginning a batch of %s, post %s of %s', $batch_size, $batch_position, $total_count->total ) );
+					}
+					foreach ( $post_processing_data as $row ) {
+						$post_id = $row->post_id;
+						$processor_type = get_post_meta( $post_id, 'tmsc_processor_type' ,true );
+
+						if ( ! empty( $processor_type ) ) {
+							$relationship_map = apply_filters( "tmsc_{$processor_type}_relationship_map", array() );
```
Comment (kjbenk) [inline inc/class-tmsc-sync.php]: Add docblock for the filter.

https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/php/#4-hooks-actions-and-filters

File: inc/class-tmsc-sync.php [hunk 3]
```diff
@@ -308,45 +324,58 @@ public function complete_sync() {
...
+							$data = maybe_unserialize( $row->meta_value );
+							foreach ( $data as $key => $ids ) {
```
Comment (kjbenk) [inline inc/class-tmsc-sync.php]: Do you need to ensure ` ! empty( $data ) && is_array( $data )` before looping through `$data`?

File: inc/class-tmsc-sync.php [hunk 4]
```diff
@@ -356,43 +385,84 @@ public function do_post_processing() {
...
-										if ( ! empty( $related_constituents ) ) {
-											$meta_data[ $role_slug ] = $related_constituents;
+
+										if ( ! empty( $related_posts ) && $existing_relationships !== $related_posts ) {
+											update_post_meta( $post_id, $key, $related_posts );
+										}
+									} elseif ( 'constituent' === $relationship_map[ $key ]['type'] ) {
+										foreach ( $ids as $constituent_type_slug => $role_data ) {
+											$meta_data = array();
+											foreach( $role_data as $role_slug => $constituent_ids ) {
```
Comment (kjbenk) [inline inc/class-tmsc-sync.php]: Do you need to ensure ` ! empty( $role_data ) && is_array( $role_data )` before looping through `$role_data`?

File: inc/class-tmsc-sync.php [hunk 5]
```diff
@@ -356,43 +385,84 @@ public function do_post_processing() {
...
+													}
+												}
+											}
+											if ( ! empty( $meta_data ) ) {
+												update_post_meta( $post_id, $constituent_type_slug, $meta_data );
+											}
+										}
+									} elseif ( 'link' === $relationship_map[ $key ]['type'] ) {
+										$link_meta = array();
+										if ( apply_filters( 'tmsc_enable_links', false ) ) {
```
Comment (kjbenk) [inline inc/class-tmsc-sync.php]: Document filter

File: inc/class-tmsc-sync.php [hunk 6]
```diff
@@ -356,43 +385,84 @@ public function do_post_processing() {
...
-									$link_meta[] = array(
-										'link' => $resource_data,
-									);
-								}
-								if ( ! empty( $link_meta ) ) {
-									update_post_meta( $post_id, $key, $link_meta );
-								}
 							}
 						}
+						if ( apply_filters( 'tmsc_delete_post_processing_data', false ) ) {
```
Comment (kjbenk) [inline inc/class-tmsc-sync.php]: Document filter

File: inc/class-tmsc-sync.php [hunk 7]
```diff
@@ -356,43 +385,84 @@ public function do_post_processing() {
...
+							delete_post_meta( $post_id, 'tmsc_post_processing' );
+						}
+					}
+
+					tmsc_stop_the_insanity();
```
Comment (kjbenk) [inline inc/class-tmsc-sync.php]: Should this function only be called in a `WP_CLI` context?

File: inc/class-tmsc-sync.php (hunk 1 above)
Comment (jaredcobb) [reply inc/class-tmsc-sync.php]: Good catch, thank you

File: inc/class-tmsc-sync.php (hunk 7 above)
Comment (jaredcobb) [reply inc/class-tmsc-sync.php]: Yes, this script may be called via a cron job and in any context we should prevent memory leaks.

**Pull request merged**